<!DOCTYPE html>
<html>
<head>
<title>README.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="singletrome">Singletrome</h1>
<p>Singletrome, is an enhanced human lncRNA genome annotation for scRNA-seq analysis, by merging protein-coding and lncRNA databases with additional filters for quality control. Using Singletrome, we were able to cluster cell types based solely on lncRNAs expression, providing evidence of the depth and diversity of lncRNA reads contained in scRNA-seq data. Our analysis identified lncRNAs differentially expressed in specific cell types with development of liver fibrosis. Importantly, lncRNAs alone were able to predict cell types and human disease pathology through the application of machine learning. This comprehensive annotation will allow mapping of lncRNA expression across cell types of the human body facilitating the development of an atlas of human lncRNAs in health and disease.</p>
<h3 id="the-scripts-for-creating-singletrome-and-for-processing-single-cell-rna-seq-10x-genomics-data-are-uploaded-and-explained-below">The scripts for creating Singletrome and for processing single cell RNA-Seq 10X Genomics Data are uploaded and explained below.</h3>
<h4 id="genomescripts">Genome_scripts</h4>
<p>The scripts in the Genome_scripts were used to download the genome fasta files, genome annotation from GENCODE and LncExpDB, scripts to build genome indices and create BED file format for comparing protein-coding and lncRNA genes as well as to perform RSEQC analysis. The gene_and_transcript_length_analysis.R has the code to perform correlation analysis of gene and transcript length for both protein-coding genes and lncRNA genes.</p>
<h4 id="gtfprocessesing">GTF_Processesing</h4>
<p>GtfProcessor.ipynb contains all the code and functions to read and write a GTF file with Genes, Transcripts and Exons. create_ULGA_and_TLGA_gtfs.ipynb contains all the code to create TLGA and ULGA genome annotations. This notebook deletes lncRNA genes that overlap protein-coding genes on the same strand and trim lncRNA exons that overlap protein-coding genes on the opposite strand.</p>
<h4 id="analysispipelines">ANALYSIS_PIPELINES</h4>
<p>The scripts in the ANALYSIS_PIPELINES were used to perform analysis and create plots such as CellrangerCompiler.Rmd to compile and compare the mapping stats produced by the cellranger, Expression_Comparison_and_Filtering.Rmd to filter lncRNAs by comparing the expression of lncRNAs in TLGA and ULGA, Read_Distribution_Filtering.Rmd filter lncRNAs by assessing the read distribution across the transcripts, SeuratPipeline.Rmd contains the analysis of protein-coding and lncRNA genes across different genome annotations and Filtered_DownStream_Analysis.R to perform clustering and differential expression of the lncRNAs that passed the quality control filter. The geneBody_coverage_modified.py is the modified script form RSEQC to report the coverage for each transcript before normalizing to 1.  CustomTheme.R is the theme that is used to plot figures.</p>
<h4 id="utility">Utility</h4>
<p>The Utility directory contains FUNCTIONS.R which has the functions to perform analysis. This script is used by many other scripts and contain FUNCTIONS that are utilized by other scripts.</p>
<h4 id="machinelearning">Machine_Learning</h4>
<p>The Machine_Learning directory contains scripts that were used to convert the Seurat objects into matrix form for cell type and disease prediction. The directory also includes the code to split data into training and testing sets as well as all the models for cell type and disease prediction.</p>
<h5 id="creatematricesformlr">Create_Matrices_For_ML.R</h5>
<p>Creates two matrices from Seurat Object, (i) 'Matrix.csv' that contains the counts and (ii) 'metadata.csv' that contains the metadata for the counts. This metadata contains additional information and the cell type or disease annotation for machine learning prediction.</p>
<h5 id="the-script-preprocessingipynb">The script PreProcessing.ipynb</h5>
<ul>
<li>Takes two files (matrices) as input, (i) 'Matrix.csv' that contains the counts and (ii) 'metadata.csv'  that contains the metadata for the counts.</li>
<li>Merge these two files (matrices) based on the cell bar code.</li>
<li>Identify the columns where all values are zero.</li>
<li>Delete the columns with zero values.</li>
<li>Create Dummy Variables for the outcome variables such as Cell Type or Disease.</li>
<li>Saves the resultant data frame to CSV.</li>
<li>The output CSV is used as input for XGBoost Classifier to predict cell type or disease.</li>
</ul>
<h5 id="celltype-prediction-classifieripynb">CellType-Prediction-Classifier.ipynb</h5>
<li> XGBoost classifier to predict cell types.</li>
<h5 id="disease-prediction-classifieripynb">Disease-Prediction-Classifier.ipynb</h5>
<li>XGBoost classifier to predict disease.</li>

</body>
</html>
